src/py/flwr/server/superlink/fleet/grpc_rere/server_interceptor.py

```diff
@@ -0,0 +1,176 @@
+from flwr.common.secure_aggregation.crypto.symmetric_encryption import (
+    bytes_to_public_key,
+    generate_shared_key,
+    public_key_to_bytes,
+    verify_hmac,
+)
+from flwr.proto.fleet_pb2 import (  # pylint: disable=E0611
+    CreateNodeRequest,
+    CreateNodeResponse,
+    DeleteNodeRequest,
+    DeleteNodeResponse,
+    GetRunRequest,
+    GetRunResponse,
+    PullTaskInsRequest,
+    PullTaskInsResponse,
+    PushTaskResRequest,
+    PushTaskResResponse,
+)
+from flwr.server.superlink.state import State
+
+_PUBLIC_KEY_HEADER = "public-key"
+_AUTH_TOKEN_HEADER = "auth-token"
+
+Request = Union[
+    CreateNodeRequest,
+    DeleteNodeRequest,
+    PullTaskInsRequest,
+    PushTaskResRequest,
+    GetRunRequest,
+]
+
+Response = Union[
+    CreateNodeResponse,
+    DeleteNodeResponse,
+    PullTaskInsResponse,
+    PushTaskResResponse,
+    GetRunResponse,
+]
+
+
+def _get_value_from_tuples(
+    key_string: str, tuples: Sequence[Tuple[str, Union[str, bytes]]]
+) -> bytes:
+    value = next((value for key, value in tuples if key == key_string), "")
+    if isinstance(value, str):
+        return value.encode()
+
+    return value
+
+
+class AuthenticateServerInterceptor(grpc.ServerInterceptor):  # type: ignore
+    """Server interceptor for client authentication."""
+
+    def __init__(
+        self,
+        state: State,
+        client_public_keys: Set[bytes],
+        private_key: ec.EllipticCurvePrivateKey,
+        public_key: ec.EllipticCurvePublicKey,
+    ):
+        self._lock = threading.Lock()
+        self.server_private_key = private_key
+        self.server_public_key = public_key
+        self.state = state
+        self.state.store_client_public_keys(client_public_keys)
+        self.encoded_server_public_key = base64.urlsafe_b64encode(
+            public_key_to_bytes(self.server_public_key)
+        )
+        log(
+            INFO,
+            "Client authentication enabled with %d known public keys",
+            len(client_public_keys),
+        )
+
+    def intercept_service(
+        self,
+        continuation: Callable[[Any], Any],
+        handler_call_details: grpc.HandlerCallDetails,
+    ) -> grpc.RpcMethodHandler:
+        """Flower server interceptor authentication logic.
+
+        Intercept all unary calls from clients and authenticate clients by validating
+        auth metadata sent by the client. Continue RPC call if client is authenticated,
+        else, terminate RPC call by setting context to abort.
+        """
+        message_handler: grpc.RpcMethodHandler = continuation(handler_call_details)
+        return self._generic_auth_unary_method_handler(message_handler)
+
+    def _generic_auth_unary_method_handler(
+        self, message_handler: grpc.RpcMethodHandler
+    ) -> grpc.RpcMethodHandler:
+        def _generic_method_handler(
+            request: Request,
+            context: grpc.ServicerContext,
+        ) -> Any:
+            with self._lock:
```

Comment:
Why do we need the lock here? The scope of the lock pretty much prevents us from handling multiple requests in parallel

Reply:
@danieljanes it manipulates the state, so we need the lock.

Reply:
Two clients maybe connecting to the server at the same time, so the state must be accessed once at a time to avoid race condition.

Reply:
But do we need to lock the entire scope, or is it sufficient to lock only those parts concerned with `CreateNodeRequest` (i.e., only the part that actually modifies state)?

Reply:
Removed due to next comment.